File: day04.py
"""Solution to day 4 for Advent of Code"""

from get_input import get_input, line_parser


def part1(ranges):
    return sum(1 for (f_start, f_end, s_start, s_end) in ranges
               if (s_start <= f_start and f_end <= s_end) or
                  (f_start <= s_start and s_end <= f_end))


def part2(ranges):
    return sum(1 for (f_start, f_end, s_start, s_end) in ranges
               if not (f_end < s_start or s_end < f_start))


def parse(line):
    return tuple(int(n) for part in line.split(',') for n in part.split('-'))


if __name__ == "__main__":
    LINES = line_parser(get_input(day=4, year=2022), parse=parse)
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")

File: day04.rs
use common::get_input;

type Input = Vec<((usize, usize), (usize, usize))>;

fn part1(ranges: &Input) -> usize {
    ranges.iter().filter(|(first, second)|
        (second.0 <= first.0 && first.1 <= second.1) ||
        (first.0 <= second.0 && second.1 <= first.1)
    ).count()
}


fn part2(ranges: &Input) -> usize {
    ranges.iter().filter(|(first, second)| !(first.1 < second.0 || second.1 < first.0)).count()
}

fn parse(text: String) -> Input {
    text.trim().split("\n").map(|line| {
        let parts = line.split(",")
            .map(|p| p.split("-").map(|n| n.parse::<usize>().unwrap())).flatten()
            .collect::<Vec<usize>>();
        assert!(parts.len() == 4);
        ((parts[0], parts[1]), (parts[2], parts[3]))
    }).collect()
}

fn main() {
    let input = parse(get_input(04, 2022));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
